Debug dump utility HLD

vivekreddynv opened the pull request:
Signed-off-by: Vivek Reddy Karri vkarri@nvidia.com

The high-level design document for the dump cli utility

| Repo  | PR title | State |  
| ---------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | 
| sonic-utilities | [[debug dump util] Match Infrastructure](https://github.com/Azure/sonic-utilities/pull/1666) | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-utilities/1666) |  
| sonic-utilities | [[debug dump util] Module implementation Logic and Port Module](https://github.com/Azure/sonic-utilities/pull/1667)  | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-utilities/1667) |   
| sonic-utilities | [[debug dump util] Base Skeleton and Click Class added](https://github.com/Azure/sonic-utilities/pull/1668) | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-utilities/1668) |  
| sonic-utilities | [[debug dump util] Techsupport addition](https://github.com/Azure/sonic-utilities/pull/1669) | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-utilities/1669) | 
| sonic-utilities | [[debug dump util] COPP Module Added](https://github.com/Azure/sonic-utilities/pull/1670) | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-utilities/1670) |

gechiang commented:
@SuvarnaMeenakshi Can you help clarify if VS for multi-asic support is available so that @vivekreddynv can add the support of this utilities for multi-asic? 
Thanks!

SuvarnaMeenakshi commented:
> @SuvarnaMeenakshi Can you help clarify if VS for multi-asic support is available so that @vivekreddynv can add the support of this utilities for multi-asic?
> Thanks!
for master branch, this PR is required to bring up multi-asic vs: https://github.com/Azure/sonic-buildimage/pull/7632
PR verification steps have information on how to bring up a multi-asic vs testbed.
Also, swss docker does not come up currently, I am debugging that part currently.
To workaround, this diff can be used for now: https://github.com/SuvarnaMeenakshi/sonic-buildimage/commit/f7f626049d06356708ce6d21012b1cdffd8255cb

vivekreddynv commented:
@gechiang I've added "--namespace / -n" option to the `dump state` CLI. I'm using `SonicV2Connector` to connect to the Redis Underneath and i've made sure that this argument is propagated down to the connector level. I've also added the Multi-ASIC UT's in the sonic-utilities directory and made sure the change is working as expected. 

However, i couldn't test this feature on actual hardware or on a VS Image currently. I'll try my best to test this on a VS image once the issues mentioned by @SuvarnaMeenakshi are merged and also when i find some buffer. 

If you wish to verify the functionality on a multi-asic platform, i can send the code tarball and instructions or a binary built with these changes to test the feature

liat-grozovik commented:
@lguohan can you please assign a reviewer to approve this HLD?
PRs should come by EOW and HLD should be closed before it. Thanks.

SuvarnaMeenakshi commented on doc/Dump-Utility.md:
```diff
@@ -0,0 +1,823 @@
+    def execute(self, param):
+        self.template = display_template(dbs=["CONFIG_DB", "APPL_DB", "ASIC_DB"])
+        port = param[ARG_NAME]
+	ns = param["namespace"]  # namespace info, if any provided by the user
```
In multi-asic, there could be some modules/features specific to default namespace. Ex: SNMP.
If we try to get SNMP data from CONFIG_DB, and pass a namespace, the key will not available in that namespace. so should it error out saying  table/key not found in the namespace provided?

vivekreddynv replied:
Yeah, that will likely happen, if not handled inside the module. 

If a module was to be programmed for SNMP and if we know, the data for SNMP will not be present in namespaces other than DEFAULT, then module implementor should take that into consideration when using MatchInfra to pull the data.

SuvarnaMeenakshi commented:
> @gechiang I've added "--namespace / -n" option to the `dump state` CLI. I'm using `SonicV2Connector` to connect to the Redis Underneath and i've made sure that this argument is propagated down to the connector level. I've also added the Multi-ASIC UT's in the sonic-utilities directory and made sure the change is working as expected.
> 
> However, i couldn't test this feature on actual hardware or on a VS Image currently. I'll try my best to test this on a VS image once the issues mentioned by @SuvarnaMeenakshi are merged and also when i find some buffer.
> 
> If you wish to verify the functionality on a multi-asic platform, i can send the code tarball and instructions or a binary built with these changes to test the feature

You could use the latest master branch to bring up multi-asic Virtual swith, the required changes are in place.
https://github.com/Azure/sonic-buildimage/blob/master/platform/vs/README.vsvm.md - Reference.

vivekreddynv commented on doc/Dump-Utility.md:
```diff
@@ -0,0 +1,823 @@
+                    }
+                }
+            ],
+            "tables_not_found": []
```
That column will have those table names, which are supposed to have the info which the module was looking for and yet it did not find one.

vivekreddynv commented on doc/Dump-Utility.md:
```diff
@@ -0,0 +1,823 @@
+}
+
+# Specify the namespace from which the keys should be fetched from using --namespace
+root@sonic# dump state port Ethernet0 --db APPL_DB --namespace asic0
```
If nothing is specified, the default namespace is used.  Yeah, unless you filter them based on --db name and if the module is programmed to look for a that, that DB name will be present in the dump output. 

So, the output will always be specific to a namespace.

vivekreddynv commented on doc/Dump-Utility.md:
```diff
@@ -0,0 +1,823 @@
+   }
+}
+Note: DB_NAME could be either of ["CONFIG_DB", "APPL_DB", "ASIC_DB", "STATE_DB", ....] or "CONFIG_FILE".
+COPP is a example which will have a entry "CONFIG_FILE". This entry is required for COPP because, the default COPP entries are read from copp_cfg.json by coppmgr and are not present in the CONFIG_DB.  
```
Currently, only JSON file supported.

vivekreddynv commented on doc/Dump-Utility.md:
````diff
@@ -0,0 +1,823 @@
+
+```
+root@sonic# dump state --help
+Usage: main.py state [OPTIONS] MODULE IDENTIFIER
````
Resolved!

SuvarnaMeenakshi commented on doc/Dump-Utility.md:
```diff
@@ -0,0 +1,823 @@
+
+### 1.1 Intended Usage
+
+Note: Currently, `state` is the only subcommand under `dump`.
```
Will 'state' be confusing to understand.
For example, if we say dump "state", will use expect to dump the "state_db"? 
Will user think that, dump could either dump state or dump config.
What is the intention of adding sub command under dump? Is the idea to use dump for any thing else in future? is there an example where dump could be used for something else in future?

vivekreddynv replied:
Yeah, that was indeed the idea, if we choose to use the dump command for something else apart from fetching just redis-state. For example, Maybe to fetch data from sdk dumps, etc.

As for the state subcommand,  the help should make it clear.

```
root@sonic# dump state --help
Usage: dump state [OPTIONS] MODULE IDENTIFIER

  Dump the redis-state of the identifier for the module specified

Options:
  -s, --show            Display Modules Available
  -d, --db TEXT         Only dump from these Databases
  -t, --table           Print in tabular format  [default: False]
  -k, --key-map         Only fetch the keys matched, don't extract field-value dumps  [default: False]
  -v, --verbose         Prints any intermediate output to stdout useful for dev & troubleshooting  [default: False]
  -n, --namespace TEXT  Dump the redis-state for this namespace.  [default: DEFAULT_NAMESPACE]
  --help                Show this message and exit.
```

SuvarnaMeenakshi replied:
Ok, The help sentence should be modified to convey that the CLI is not just to dump redis-state but also can dump from a CONFIG_FILE.
Also, is there any help in the command that tells the user that -d/--db option can take db name or CONFIG_FILE?

vivekreddynv replied:
1) Yeah, will add that
2) Currently, there is no way to notify the user about the DB/CONFIG_FILE's that the particular module will be looking, directly from the CLI. Can add this to the roadmap for next revision though.

SuvarnaMeenakshi replied:
Ok, It is better to update the help message for --db sub command, telling "Dump from specified database or config_file".

vivekreddynv replied:
Updated them in the HLD.

SuvarnaMeenakshi replied:
Help message of "dump state" could be shortened as:
 "Dump the current state of the identifier for the specified module from Redis DB or CONFIG_FILE"

vivekreddynv replied:
Done

liat-grozovik commented:
@SuvarnaMeenakshi kindly reminder to review recent updates

SuvarnaMeenakshi reviewed (approved):
LGTM
